### PR TITLE
Remove atomics that are not needed or flat out wrong

### LINKED ISSUE #766: Needs to revisit calls to atomic_xxx APIs in driver
In our drivers, we call atomic_xxx() APIs in multiple places. After examination of a few cases, I can see that some of the usage of these APIs are problematic. I'd like to list some good and not-so-good use cases that I know of. We want to double check all cases in our driver to make sure we don't have lurking bugs by calling atomic kernel APIs.

1. In our code, we sometimes use mutex together with atomic API. This is OK, but not efficient. We want to use one or the other, not both. We prefer atomic API to mutex, if we *have to* use one.

2. We don't want to call atomic API, if not necessary. Atomic API is to be used when the operand may be accessed at the same time by more than one thread at any point of time. This is to avoid half updated variable. We need to use atomic API in the above case *and* if
a) the address of the variable is NOT naturally aligned.
b) the operation cannot be done atomically to a memory location.
Case a) usually is resulted from casting one data structure to a different one. E.g., you cast a char array to a data structure. In this case, the data structure has higher alignment requirement than char array and compiler does not see that cast coming. Case b) may be a inc/dec operation, which needs to perform read-modify-write action. E.g., you may need to call atomic_inc() to increment a counter. The compiler should be able to properly align an explicitly defined (vs cast one) data structure, so there is no need to call atomic API to access it's members. The kernel memory allocator should hand out 64-bit aligned address so that the buffer allocated is suitable for any data structure. So, again, no need to use atomic API for dynamically allocated data structures.

3. In below code, you should NOT use atomic API, but need a mutex:
if (atomic_read(xxx))
    do_something();
The atomic API does not make the check and do_something an atomic operation automatically. A mutex is needed to wrap around them to make this multi-thread safe. It maybe that the race may never happen based on where these code are used. But, instead of spending time analyzing each individual case, we should employ a mutex and be done with it. This is going to be safer and the code will be easy to maintain in the future.

### PR DESCRIPTION
**Pull request by stsoe**
Fixes #766

### CODE REVIEW

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c [hunk 1]
```diff
@@ -214,11 +220,6 @@ int xocl_ctx_ioctl(struct drm_device *dev, void *data,
 		// CU unlocked explicitly
 		--client->num_cus;
 		--xdev->ip_reference[args->cu_index];
-		if (!client->num_cus) {
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: Why not unlock here if you removed the last context?

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c [hunk 2]
```diff
@@ -35,27 +35,23 @@ xuid_t uuid_null = NULL_UUID_LE;
 #endif
 
 static int
-xocl_client_lock_bitstream(struct xocl_dev *xdev, struct client_ctx *client)
-{
-	if (atomic_read(&client->xclbin_locked))
+xocl_client_lock_bitstream_helper(struct xocl_dev *xdev, struct client_ctx *client) {
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: _helper() -> _nolock() to better align the intention of this function?

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c [hunk 3]
```diff
@@ -98,8 +104,8 @@ int xocl_execbuf_ioctl(struct drm_device *dev,
 	int numdeps;
 	int ret = 0;
 
-	if (atomic_read(&xdev->needs_reset)) {
-		userpf_err(xdev, "device needs reset, use 'xbsak reset -h'");
+	if (xdev->needs_reset) {
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: Don't need a lock?

Comment (stsoe) [reply src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: You mentioned that if the data is aligned it by default atomic.  This is a Boolean only.

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c (hunk 2 above)
Comment (stsoe) [reply src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: Yes.

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c (hunk 3 above)
Comment (maxzhen) [reply src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: This is the 3rd case mentioned in the issue: behavior changed based on value of a variable shared between multiple threads. This usually indicates that a lock is needed. It was not clear to me why we don't use a lock here. Let's have some more discussion when I'm in office this afternoon.

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c (hunk 1 above)
Comment (stsoe) [reply src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c]: yes

**Comment by maxzhen** [comment]
chat with Soren, changes look good.

**Comment by stsoe** [comment]
Replacing with another PR

**Pull request closed without merging**